### PR TITLE
feat: Support Platform Routing host annotations in Authorization

### PR DESCRIPTION
**Pull request by aslakknutsen**
Platform Routing will write back Public and External addresses to the target CR and Platform Authorization needs to take these into account when setting up the AuthConfig as well as the hosts exposed in status by the target Controller.

https://issues.redhat.com/browse/RHOAIENG-11334

### CODE REVIEW

File: controllers/routing/routing_controller.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ type PlatformRoutingController struct {
 // +kubebuilder:rbac:groups="route.openshift.io",resources=routes,verbs=*
 // +kubebuilder:rbac:groups="networking.istio.io",resources=virtualservices,verbs=*
 // +kubebuilder:rbac:groups="networking.istio.io",resources=gateways,verbs=*
-// +kubebuilder:rbac:groups="networking.istio.io",resources=destinationrule,verbs=*
+// +kubebuilder:rbac:groups="networking.istio.io",resources=destinationrules,verbs=*
```
Comment (bartoszmajsak) [inline controllers/routing/routing_controller.go]: I think it deserves its own fix PR.

File: pkg/authorization/authconfig.go [hunk 2]
```diff
@@ -132,92 +133,103 @@ func (k *annotationAuthTypeDetector) Detect(_ context.Context, res *unstructured
...
 
-		hosts = parsedHosts
+		return keys
 	}
 
-	u := unique(hosts)
-	if len(u) == 0 {
-		return []string{"unknown.host.com"}, nil
+	isURI := func(host string) bool {
+		return strings.HasPrefix(host, "http://") || strings.HasPrefix(host, "https://")
```
Comment (bartoszmajsak) [inline pkg/authorization/authconfig.go]: Little nit, but maybe we should be more accurate calling it `isHTTP` instead?

File: pkg/authorization/authconfig.go [hunk 3]
```diff
@@ -132,92 +133,103 @@ func (k *annotationAuthTypeDetector) Detect(_ context.Context, res *unstructured
...
 }
 
-type expressionHostExtractor struct {
-	paths []string
-}
-
-func NewExpressionHostExtractor(paths []string) spi.HostExtractor {
-	return &expressionHostExtractor{paths: paths}
-}
+func UnifiedHostExtractor(extractors ...spi.HostExtractor) spi.HostExtractor { //nolint:gocognit //reason Inlined functions to avoid surface pollution
```
Comment (bartoszmajsak) [inline pkg/authorization/authconfig.go]: What do you mean by "surface pollution"?

File: pkg/authorization/authconfig.go [hunk 4]
```diff
@@ -132,92 +133,103 @@ func (k *annotationAuthTypeDetector) Detect(_ context.Context, res *unstructured
...
+	unique := func(in []string) []string {
```
Comment (bartoszmajsak) [inline pkg/authorization/authconfig.go:137]: Small thing, but those names helped quickly mapped out what is going on:

```golang
unique := func(in []string) []string {
		set := map[string]bool{}

		for _, elem := range in {
			set[elem] = true
		}

		unique := make([]string, len(set))

		i := 0

		for elem := range set {
			unique[i] = elem
			i++
		}

		return unique
	}
```

File: pkg/authorization/authconfig.go [hunk 5]
```diff
@@ -132,92 +133,103 @@ func (k *annotationAuthTypeDetector) Detect(_ context.Context, res *unstructured
...
 }
 
-func appendHosts(hosts []string, foundHosts ...string) ([]string, error) {
-	for _, foundHost := range foundHosts {
-		if isURI(foundHost) {
-			parsedURL, errParse := url.Parse(foundHost)
-			if errParse != nil {
-				return nil, fmt.Errorf("failed to parse URL %s: %w", foundHost, errParse)
-			}
+func NewExpressionHostExtractor(paths []string) spi.HostExtractor {
```
Comment (bartoszmajsak) [inline pkg/authorization/authconfig.go]: I wonder if `NewPathExpressionExtractor` wouldn't be more meaningful. Plus little godoc could help.

File: pkg/authorization/authconfig_test.go [hunk 6]
```diff
@@ -40,13 +40,29 @@ var _ = Describe("AuthConfig functions", test.Unit(), func() {
...
 		})
 
+		It("should return unique list", func() {
+
+			// given
+			extractor := authorization.NewExpressionHostExtractor([]string{"status.url"})
+			target := unstructured.Unstructured{
+				Object: map[string]interface{}{},
+			}
+			unstructured.SetNestedStringSlice(target.Object, []string{"test.com", "http://test.com"}, "status", "url")
```
Comment (bartoszmajsak) [inline pkg/authorization/authconfig_test.go]:
```suggestion
			unstructured.SetNestedStringSlice(target.Object, []string{"test.com", "http://test.com", "https://test.com"}, "status", "url")
```

File: pkg/authorization/authconfig_test.go [hunk 7]
```diff
@@ -40,13 +40,29 @@ var _ = Describe("AuthConfig functions", test.Unit(), func() {
...
+
+			// when
+			hosts, err := authorization.UnifiedHostExtractor(extractor)(&target)
+			// then
+			Expect(err).To(Not(HaveOccurred()))
+			Expect(hosts).To(ContainElements("test.com"))
```
Comment (bartoszmajsak) [inline pkg/authorization/authconfig_test.go]:
```suggestion
			Expect(hosts).To(HaveExactElements("test.com"))
```

File: pkg/authorization/authconfig.go [hunk 8]
```diff
@@ -132,92 +133,103 @@ func (k *annotationAuthTypeDetector) Detect(_ context.Context, res *unstructured
...
 
-func unique(in []string) []string {
-	store := map[string]bool{}
+		extractedHosts := []string{}
 
-	for _, v := range in {
-		store[v] = true
-	}
+		for _, extractor := range extractors {
+			extractedDelegateHosts, err := extractor(target)
```
Comment (bartoszmajsak) [inline pkg/authorization/authconfig.go]: What do you mean by `Delegate` here?

File: controllers/authorization/authorization_controller.go [hunk 9]
```diff
@@ -32,10 +33,13 @@ func NewPlatformAuthorizationController(cli client.Client, log logr.Logger,
...
 			"component", component.ObjectReference.Kind,
 		),
-		config:         config,
-		authComponent:  component,
-		typeDetector:   authorization.NewAnnotationAuthTypeDetector(metadata.Annotations.AuthEnabled),
-		hostExtractor:  authorization.NewExpressionHostExtractor(component.HostPaths),
+		config:        config,
+		authComponent: component,
+		typeDetector:  authorization.NewAnnotationAuthTypeDetector(metadata.Annotations.AuthEnabled),
+		// TODO: Evaluate passing in the hostExtractor to avoid coupling the authorizaiton/routing packages
```
Comment (bartoszmajsak) [inline controllers/authorization/authorization_controller.go:39]: I agree with `TODO` - wondering if we shouldn't start converting those to some refactoring story in JIRA to they won't get lost.

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
Couple of questions, few nits and actionable suggestions :)

File: controllers/routing/routing_controller.go (hunk 1 above)
Comment (aslakknutsen) [reply controllers/routing/routing_controller.go]: Oh yea that just got sucked in it seems.

File: pkg/authorization/authconfig.go (hunk 2 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go]: Well, the intent is to determine if we can call url.Parse which require a protocol. So in that sense a better name would have been isURL(in hindsight..). We currently only "know of" http/http and it's a bit of a short cut to a proper check you can say.

File: pkg/authorization/authconfig.go (hunk 3 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go]: random minor single use functions all over. "function protected scope" :)

File: pkg/authorization/authconfig.go (hunk 4 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go:137]: fixed https://github.com/maistra/odh-platform/pull/49/commits/7d94a7802a37cac651c1c3f32cef61e25e3840ce

File: pkg/authorization/authconfig.go (hunk 8 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go]: Delegate is wrong, but I couldn't think of a better name last night. Decorated maybe. 

UnifiedHostExtractor doesn't do any extraction it self, it's a wrapper to make a unique and unified view of multiple other HostExtractors. SubExtractor? RealExtractor? I don't know

File: pkg/authorization/authconfig.go (hunk 5 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go]: fixed https://github.com/maistra/odh-platform/pull/49/commits/7cf23bd8829904328a2f2c2990ca91070a186ae6

File: controllers/routing/routing_controller.go (hunk 1 above)
Comment (aslakknutsen) [reply controllers/routing/routing_controller.go]: reverted https://github.com/maistra/odh-platform/pull/49/commits/a34fec89faca6b5c0f87bbf6e04de6f2e6ec2b9e

Comment (bartoszmajsak) [reply controllers/routing/routing_controller.go]: Do a PR :)

File: pkg/authorization/authconfig.go (hunk 2 above)
Comment (bartoszmajsak) [reply pkg/authorization/authconfig.go]: Fair enough, how about `isSupportedURL` then?

File: pkg/authorization/authconfig.go (hunk 3 above)
Comment (bartoszmajsak) [reply pkg/authorization/authconfig.go]: Can you rephrase the comment based on ^? That would help me in the future to not get confused :)

File: pkg/authorization/authconfig.go (hunk 8 above)
Comment (bartoszmajsak) [reply pkg/authorization/authconfig.go]: Or `extractedHosts` -> `combinedExtractedHosts` and then `extractedDelegateHosts` -> `extractedHosts`?

File: pkg/authorization/authconfig.go (hunk 2 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go]: made it isURL. The intent of the whole thing is to extract Hosts. If it's a URL we need to parse it to find the host. SupportedURL sounds more like we're looking for a type of URL which is not the case, even tho the impl is a bit too specific to http atm. The function should be named after it's intent not the current impl. 

https://github.com/maistra/odh-platform/pull/49/commits/f7fc2725c6fc435b15ce06106a2864e0e5ee4b52

File: pkg/authorization/authconfig.go (hunk 3 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go]: https://github.com/maistra/odh-platform/pull/49/commits/15e2cadd7c36e1e831278f6667a447c4d23ff9bf

File: pkg/authorization/authconfig.go (hunk 8 above)
Comment (aslakknutsen) [reply pkg/authorization/authconfig.go]: A yes, easier way of looking at it

https://github.com/maistra/odh-platform/pull/49/commits/bd7965f4ce7558637e52e07733d317ed74ada0ea

**Review by bartoszmajsak — APPROVED** [review]
merge
